pallets/parachain-system/src/lib.rs

```diff
@@ -282,6 +295,40 @@ pub mod pallet {
 					.map(|(recipient, data)| OutboundHrmpMessage { recipient, data })
 					.collect::<Vec<_>>();
 
+			if MaxUnincludedLen::<T>::get().filter(|len| !len.is_zero()).is_some() {
+				let dmp_remaining_messages = PendingDownwardMessages::<T>::get().len() as u32;
+				let limits =
+					TotalBandwidthLimits::new(&relevant_messaging_state, dmp_remaining_messages);
+
+				let hrmp_outgoing = outbound_messages
+					.iter()
+					.map(|msg| {
+						(
+							msg.recipient,
+							HrmpChannelUpdate { msg_count: 1, total_bytes: msg.data.len() as u32 },
+						)
+					})
+					.collect();
+				let dmp_processed_count = ProcessedDownwardMessages::<T>::get();
+				let used_bandwidth = UsedBandwidth {
+					ump_msg_count,
+					ump_total_bytes,
+					hrmp_outgoing,
+					dmp_processed_count,
+				};
+				// The bandwidth constructed was ensured to satisfy relay chain constraints.
+				let ancestor = Ancestor::new_unchecked(used_bandwidth, todo!());
+
+				// Check in `on_initialize` guarantees there's space for this block.
+				UnincludedSegment::<T>::append(ancestor);
+
+				let watermark = HrmpWatermark::<T>::get();
+				AggregatedUnincludedSegment::<T>::mutate(|agg| {
+					let agg = agg.get_or_insert_with(SegmentTracker::default);
+					agg.append(&ancestor, watermark, &limits)
+						.expect("unincluded segment limits exceeded");
```

Comment:
Add a comment here to indicate that this bails because the block is invalid.

Reply:
There are checks that user code that queues HRMP messages doesn't overrun this limit in a friendly way, right?

user code should never encounter this panic if possible.

cc @ggwpez

Reply:
(ah, this is for segment length)

Reply:
I'm honestly not sure, the code performs the check w.r.t. to host config for current block, so I can imagine this bailing if the unincluded chain becomes long enough.

Reply:
@rphmeier 
```rust
/// A means of figuring out what outbound XCMP messages should be being sent.
pub trait XcmpMessageSource {
	/// Take a single XCMP message from the queue for the given `dest`, if one exists.
	fn take_outbound_messages(maximum_channels: usize) -> Vec<(ParaId, Vec<u8>)>;
}
```
I believe this trait not only should account for max messages taken from the storage, but also for segment limitations if we want to make this panic correct.
Probably primitives for segment should be extracted into common crate to be accessible from both pallets.

Reply:
Thanks. Can you file an issue about that?

Reply:
>cc @ggwpez

I have not worked with the XCMP outbound side yet, so cant help you here.